### PR TITLE
Recommendation engine optimization

### PR DESCRIPTION
**Pull request by aradoje**

### CODE REVIEW

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  image: "images/posts/neo4j.jpg"
+  teaser: images/posts/neo4j.jpg
+categories:
+- Symfony
+- Neo4j
+- Recommendation engine
+--- 
+
+In today applications recommendations became quite necessary for better audience/customer targeting. So, as you could assume, we have our own recommendation engine for that, 
+using Symfony and Neo4j as a database. Communication between the main application and the recommendation engine is done via messages sent to/consumed from RabbitMQ.
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: "Symfony with a Neo4j database"

Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: "done via messages over RabbitMQ"

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  teaser: images/posts/neo4j.jpg
+categories:
+- Symfony
+- Neo4j
+- Recommendation engine
+--- 
+
+In today applications recommendations became quite necessary for better audience/customer targeting. So, as you could assume, we have our own recommendation engine for that, 
+using Symfony and Neo4j as a database. Communication between the main application and the recommendation engine is done via messages sent to/consumed from RabbitMQ. 
+We are sending some interesting events that happened on our website to the recommendation engine, and also pulling some suggestions from it.
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: "We are sending interesting"?

Comment (Nyholm) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: "and then pulling some recommendations from it."

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+##The problem
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]:
```
## The problem
```

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+Over time we noticed something – our message queue got quite clogged, over 2 millions of messages were waiting to be written in the database! So we dug up a bit and found out 
+that Neo4j is not quite optimized for one by one writing, it is recalculating relations after each insert, and we had large amounts of messages that were just coming and coming
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: "large amounts of messages that were just coming and coming in"

The language sounds a bit strange?

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+We are sending some interesting events that happened on our website to the recommendation engine, and also pulling some suggestions from it.
+
+##The problem
+
+Over time we noticed something – our message queue got quite clogged, over 2 millions of messages were waiting to be written in the database! So we dug up a bit and found out 
+that Neo4j is not quite optimized for one by one writing, it is recalculating relations after each insert, and we had large amounts of messages that were just coming and coming 
+in and waiting to be processed. So we needed to fix that, 
+in order to have up to date recommendations for our system.
+
+##The investigation
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]:
```
## The
```

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+After reading some articles, and Neo4j official docs, we found several guidelines:
+- don’t use the ORM
+- use [UNWIND](https://neo4j.com/docs/cypher-manual/current/clauses/unwind/)   
+
+We decided to check how that really impacts the write queries to the database so we made a little experiment. We ran the insert of 500 entries with ORM, without ORM and using UNWIND:
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Use 5000 entries instead. It blows up the numbers more =)

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+ 500/500 [============================] 100%
+Started in 28 seconds.
+Starting writing 500 entries with connection.
+ 500/500 [============================] 100%
+Finished in 5 seconds.
+Started writing 500 entries with connection and UNWIND.
+Finished in 0 seconds.
+{% endhighlight %}
+
+The write with ORM took 28, using CYPHER 5, using UNWIND took below 1s!
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: 28s

5s

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+##The solution
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]:
```
## The
```

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    public function __construct(MessageBusInterface $commandBus, LoggerInterface $logger)
+    {
+        $this->commandBus = $commandBus;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param AMQPMessage[] $messages
+     *
+     * @return array|bool
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Remove

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+            return true;
+        } catch (\Throwable $e) {
+            $this->logger->log('error', $e->getMessage(), ['exception' => $e]);
+            return false;
+        } 
+    }
+}
+{% endhighlight %}
+
+registered class as a service:
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Im not sure this code example is needed.

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            ...
+        advert_served_to_user:
+            url: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%'
+            lazy:     true
+            connection_timeout: 3
+            read_write_timeout: 3
+            # requires php-amqplib v2.4.1+ and PHP5.4+
+            keepalive: false
+            # requires php-amqplib v2.4.1+
+            heartbeat: 0
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: I would remove the parameters here that is not needed

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    consumers:
+        ...
+    	batch_consumers:
+        advert_served_to_user:
+            connection:       advert_served_to_user
+            exchange_options: {name: 'advert-served-to-user', type: fanout }
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Isnt type: direct better?

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+{% highlight php %}
+<?php
+
+namespace App\Message\CommandHandler\Advert;
+
+use App\Message\Command\Advert\RunAdvertServedToUserBatch;
+use App\Message\CommandHandler\BaseCommandHandler;
+use App\Model\AdvertServedToUserM
+
+class RunAdvertServedToUserBatchHandler extends BaseCommandHandler
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: We could (for the sake of the example) move this code to the consumer. That may make it easier to understand. What do you think=

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md (hunk 6 above)
Comment (aradoje) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: yea, I thought that and then I did not want to wait that long :D

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md (hunk 10 above)
Comment (aradoje) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: we are using service id in  `old_sound_rabbit_mq` `consumer` as a `callback` parameter

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md (hunk 11 above)
Comment (aradoje) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Which ones?

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md (hunk 13 above)
Comment (aradoje) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: sure

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md (hunk 11 above)
Comment (Nyholm) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: lazy, connection_timeout, read_write_timeout, keepalive, heartbeat

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+            url: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%'
+            lazy:     true
+            connection_timeout: 3
+            read_write_timeout: 3
+            # requires php-amqplib v2.4.1+ and PHP5.4+
+            keepalive: false
+            # requires php-amqplib v2.4.1+
+            heartbeat: 0
+    consumers:
+        ...
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]:
```
# ...
```

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+{% endhighlight %}
+
+and use our consumer as callback for batch RabbitMQ
+
+{% highlight yaml %}
+#config/packages/old_sound_rabbit_mq.yaml
+old_sound_rabbit_mq:
+    connections:
+        default:
+            ...
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]:
```
# ...
```

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+    	batch_consumers:
+            advert_served_to_user:
+                connection:       advert_served_to_user
+                exchange_options: {name: 'advert-served-to-user', type: fanout }
+                queue_options:    {name: 'advert-served-to-user'}
+                callback:         'App\Consumer\AdvertServedToUserToUserBatchConsumer'
+                qos_options:      {prefetch_size: 0, prefetch_count: 5000, global: false}
+                timeout_wait:     5
+                auto_setup_fabric: false
+                idle_timeout_exit_code: -2
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Remove

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+        ...
+    	batch_consumers:
+            advert_served_to_user:
+                connection:       advert_served_to_user
+                exchange_options: {name: 'advert-served-to-user', type: fanout }
+                queue_options:    {name: 'advert-served-to-user'}
+                callback:         'App\Consumer\AdvertServedToUserToUserBatchConsumer'
+                qos_options:      {prefetch_size: 0, prefetch_count: 5000, global: false}
+                timeout_wait:     5
+                auto_setup_fabric: false
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Remove

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+                idle_timeout_exit_code: -2
+                keep_alive: false
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Remove

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+                graceful_max_execution:
+                    timeout: 60
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: Remove

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+Finished in 2 minutes and 21 seconds.
+Started writing 5000 entries with CYPHER.
+ 5000/5000 [============================] 100%
+Finished in 0 minutes and 43 seconds.
+Started writing 5000 entries with CYPHER using UNWIND.
+Finished in 0 minutes and 4 seconds.
+
+{% endhighlight %}
+
+The write with ORM took 2 minutes and 21 second, using CYPHER 43 seconds, using UNWIND only 4 seconds!
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: s/second/seconds/

**Comment by aradoje** [comment]
Thank you.

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md (hunk 12 above)
Comment (aradoje) [reply _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]: I can switch it.

File: _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+
+We had to change routing for the specific message in order to put it on the different queue:
+
+{% highlight yaml %}
+#config/packages/messenger.yaml
+framework:
+    messenger:
+        transports:
+            batch_messages_amqp: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%/advert-served-to-user
+        ...
```
Comment (Nyholm) [inline _posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md]:
```
# ...
```

**Pull request merged**